frontend/src/metabase-lib/lib/Dimension.ts

```diff
@@ -835,120 +810,55 @@ export class FieldDimension extends Dimension {
+      name: this.isStringFieldName() && this.fieldIdOrName(),
+      display_name: this.fieldIdOrName(),
+      base_type: this.getOption("base-type"),
     });
   }
 
-  _combineFieldWithExtraMetadata(field: Field | undefined, fieldMetadata) {
-    if (field) {
-      if (!fieldMetadata || field === fieldMetadata) {
-        return field;
-      }
-
-      const fieldObject = merge(
-        field instanceof Field ? field.getPlainObject() : field,
-        fieldMetadata instanceof Field
-          ? fieldMetadata.getPlainObject()
-          : fieldMetadata,
-      );
-      return this._createField(fieldObject, { hydrate: true });
-    }
-
-    if (fieldMetadata) {
-      if (fieldMetadata instanceof Field) {
-        return fieldMetadata;
-      }
-
-      return this._createField(fieldMetadata);
-    }
-  }
-
   field(): Field {
     // If a Field is cached on the FieldDimension instance, we can shortwire this method and
     // return the cached Field.
-    const cachedField = this._getTrustedFieldCachedOnInstance();
-    if (cachedField) {
-      return cachedField;
-    }
-
-    let fieldMetadata;
-
-    // Models contain custom metadata for fields, but when these fields have integer ids,
-    // they are clobbered by the metadata from the real table that the card is based on,
-    // so we need to check the result_metadata of the nested card and merge it with the
-    // Field that exists in the Metadata object.
-    if (this.isIntegerFieldId()) {
-      fieldMetadata =
-        this._getFieldMetadataFromSavedDataset() ||
-        this._getFieldMetadataFromNestedDataset();
-    }
-
-    // In scenarios where the Field id is not an integer, we need to grab the Field from the
-    // query's table, because the Field won't exist in the Metadata object (and string Field ids
-    // are not sufficiently unique to be stored properly in the Metadata object).
-    fieldMetadata = fieldMetadata || this._getFieldMetadataFromQueryTable();
-
-    // The `fieldMetadata` object may have metadata that overrides the regular field object
-    // (e.g. a custom field display name or description on a model)
-    const field = this._metadata?.field(this.fieldIdOrName());
-    const combinedField = this._combineFieldWithExtraMetadata(
-      field,
-      fieldMetadata,
-    );
+    const locallyCachedField = this._getTrustedFieldCachedOnInstance();
+    if (locallyCachedField) {
+      return locallyCachedField;
+    }
+
+    // Prioritize pulling a `field` from the Dimenion's associated query (if one exists)
+    // because it might have locally overriding metadata on it.
+    const fieldFromQuery = this._findMatchingQueryField();
+    if (fieldFromQuery) {
+      return fieldFromQuery;
+    }
 
-    if (combinedField) {
-      return combinedField;
+    const maybeTableId = this._query?.table()?.id;
+    const fieldFromGlobalState =
+      this._metadata?.field(this.fieldIdOrName(), maybeTableId) ||
+      this._metadata?.field(this.fieldIdOrName());
```

Comment:
Not a huge fan of this line, but it seems that my approach with looking at the query's fields is not wholly sufficient for the fields of joins, and adding this fallback fixes things, I think. I'm going to try to unravel that in a separate PR to follow this one.